install_script.md

```diff
@@ -0,0 +1,154 @@
+# Install script
+
+An install script is a fragment of data which describes how to install a mod.
+
+Each version of a project can have a different install script.
+
+An install script consists of a *list* of steps referencing the files in the version.
+
+Example:
+`[{"type": "placeInDirectory", "directory": "mods", "file": "ExampleMod-1.0.0.jar"}, {"type": "placeInDirectory", "directory": "coremods", "file": "ExampleMod-1.0.0-coremod.jar"}]`
+
+## `placeInDirectory` step
+
+Download a file into a subdirectory of the Minecraft instance directory.
+
+Example:
+
+`{"type": "placeInDirectory", "directory": "mods", "file": "ExampleMod-1.0.0.jar"}`
+
+`directory` may be a nested subdirectory, e.g. `"mods/1.7.10"`. Any parent directories which don't already exist are created.
+`directory` may not contain a `.` or `..` part, unless the entire value of `directory` is `"."` (which places the file in the instance directory).
+
+## `extractZip` step
+
+Download a file *and extract it* into a subdirectory of the Minecraft instance directory.
+
+Example:
+
+`{"type": "extractZip", "directory": "config", "file": "modpack-configs.zip"}`
+
+`directory` may be a nested subdirectory, e.g. `"mods/1.7.10"`. Any parent directories which don't already exist are created.
+`directory` may not contain a `.` or `..` part, unless the entire value of `directory` is `"."` (which extracts the file into the instance directory).
+
+## `jarMod` step
+
+Download a file and merge it into `minecraft.jar` (or `minecraft-server.jar`), overwriting files that already exist.
+
+Note: this is considered an obsolete installation technique. Newer loaders typically add themselves to the classpath instead.
+
+`before` and `after` allow this project's jar mod to be prioritized when multiple jar mods are present. If a mod mentioned in `before` or `after`
+is not installed, this does not cause it to be installed - the entry does nothing.
+
+Example:
+`{"type": "jarMod", "file": "modloadermp-1.0.0.jar", "before": ["minecraft-forge"], "after": ["modloader"]}`
+
+## `addClasspathLibrary` step
+
+Download a file and add it to the classpath.
+
+Inside the "library" field is an object identical to Minecraft's version.json library specification,
+except that wherever there is a `sha1`/`size`/`url` triplet in Mojang's format, the install script
+MAY instead contain a `modip_filename` which refers to one of the version's files.
```

Comment:
This is a good idea, but right now in the spec files only specify an SHA256 hash which doesn’t work for the triplet.

Reply:
You mean the filename is optional? It could be a sha1 then - every time you see a sha1 compare it to the files list.

Reply:
by the way sha1 was broken. It should really be sha256 throughout the entire spec.

Reply:
Of course it can. The launcher can just calculate the hash of the file. It's also irrelevant because the launcher probably manages the classpath itself.

Reply:
Got it. In that case, it looks good.